Add Parametric Soft Exponential Unit (PSEU) activation layer

SriRangaTarun opened the pull request:
<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/keras-team/keras-contrib/blob/master/CONTRIBUTING.md
-->

**- What I did**
See above

**- How I did it**


**- How you can verify it**
I added unit tests

<!-- 
You need a good justification for not including tests if your pull request is 
a bug fix or adds a new feature to Keras-contrib. 
-->



---
This pull request fixes #issue_number_here

SriRangaTarun commented:
@gabrieldemarmiesse I have added the Parametric Soft Exponential Unit (PSEU) activation layer. Here is the [research paper on arXiv](https://arxiv.org/pdf/1602.01321.pdf). The build is passing now. Do you know someone who can review this ?

gabrieldemarmiesse commented:
@SriRangaTarun We're currently focused on getting the design of this repo right to have a v1.0. So there might be delays to reviews PRs which add new features.

RaphaelMeudec commented on keras_contrib/layers/advanced_activations/pseu.py:
```diff
@@ -0,0 +1,87 @@
+# -*- coding: utf-8 -*-
+from keras import initializers
```
Usually, we like to have the imports sorted alphabetically, with first all the `import XXX` and then all the `from XX import YY`. See [this StackOverflow thread](https://stackoverflow.com/questions/20762662/whats-the-correct-way-to-sort-python-import-x-and-from-x-import-y-statement)

SriRangaTarun replied:
Fixed

RaphaelMeudec commented:
@SriRangaTarun Thanks for the contribution! I've left some comments. Also, as soon as https://github.com/keras-team/keras-contrib/pull/441 gets merged, you can rebase on it and add yourself as the code owner for this feature. Nice work!

SriRangaTarun commented on keras_contrib/layers/advanced_activations/pseu.py:
```diff
@@ -0,0 +1,87 @@
+
+        self.build = True
+
+    def call(self, x, mask=None):
```
Removed. It has no impact.

SriRangaTarun commented on keras_contrib/layers/advanced_activations/pseu.py:
```diff
@@ -0,0 +1,87 @@
+        self.build = True
+
+    def call(self, x, mask=None):
+        if self.alpha_init is not None and self.alpha_init < 0:
```
Removed.

SriRangaTarun commented on keras_contrib/layers/__init__.py:
```diff
@@ -4,6 +4,7 @@
 from .advanced_activations.srelu import SReLU
 from .advanced_activations.swish import Swish
 from .advanced_activations.sinerelu import SineReLU
+from .advanced_activations.pseu import PSEU
```
Fixed.

SriRangaTarun commented:
@RaphaelMeudec I have made all the changes you suggested and the CI build is still passing.

RaphaelMeudec commented:
@gabrieldemarmiesse PR looks good to me. You might want to add @SriRangaTarun in your current codeowner PR?

SriRangaTarun commented:
@RaphaelMeudec and @gabrieldemarmiesse Everything is working fine. I have added important parameters and handled all possible input cases. But, since [the codeowners PR](https://github.com/keras-team/keras-contrib/pull/441) has been merged, it is failing the codeowners test.

SriRangaTarun commented:
@RaphaelMeudec and @gabrieldemarmiesse Today I tried to add initializer and trainable options, instead of only having a constant alpha initializer. But, I removed them because they cause numerical instability in the function (it is a piecewise function in alpha, so it becomes unstable when the negative formula is used for positive alpha weights and the other way round). A custom initializer or trainability can cause the signs of the weights to mismatch the function and result in NaNs.  So, I am sticking to constant initial alpha weights which can not be trained.

SriRangaTarun commented:
@RaphaelMeudec and @gabrieldemarmiesse Everything seems to work fine. But, ever since [the codeowners PR](https://github.com/keras-team/keras-contrib/pull/441) has been merged, it is failing the codeowners test.

SriRangaTarun commented:
@RaphaelMeudec and @gabrieldemarmiesse Today I tried to add initializer and trainable options, instead of only having a constant alpha initializer. But, I removed them because they cause numerical instability in the function (it is a piecewise function in alpha, so it becomes unstable when the negative formula is used for positive alpha weights and the other way round). A custom initializer or trainability can cause the signs of the weights to mismatch the function and result in NaNs. So, I am sticking to constant initial alpha weights which can not be trained.

SriRangaTarun commented:
@RaphaelMeudec and @gabrieldemarmiesse Today I tried to add initializer and trainable options, instead of only having a constant alpha initializer. But, I removed them because they cause numerical instability in the function (it is a piecewise function in alpha, so it becomes unstable when the negative formula is used for positive alpha weights and the other way round). A custom initializer or trainability can cause the signs of the weights to mismatch the function and result in NaNs. So, I am sticking to constant initial alpha weights which can not be trained.

SriRangaTarun commented:
Now the build is passing after adding pseu.py to CODEOWNERS.

SriRangaTarun commented:
@RaphaelMeudec and @gabrieldemarmiesse Today I tried to add initializer and trainable options, instead of only having a constant alpha initializer. But, I removed them and reverted to the previous version because they cause numerical instability in the function (it is a piecewise function in alpha, so it becomes unstable when the negative formula is used for positive alpha weights and the other way round). A custom initializer or trainability can cause the signs of the weights to mismatch the function and result in NaNs. So, I am sticking to constant initial alpha weights which can not be trained.

SriRangaTarun commented:
@RaphaelMeudec Do you suggest any other changes ?

SriRangaTarun commented:
@RaphaelMeudec Do you suggest any other changes ? The most recent build failed because the instancenormalization test suddenly failed for some reason.

RaphaelMeudec commented:
@SriRangaTarun All good, nothing to add IMO!

@gabrieldemarmiesse Failing test doesn't seem linked to this PR, code looks good to me!

SriRangaTarun commented:
Somehow, the build is passing now.

RaphaelMeudec commented:
@gabrieldemarmiesse I think we can merge this!

gabrieldemarmiesse commented:
We'll merge this after I merge the tf.keras PR. It's getting quite painful to resolve conflicts every two days.

gabrieldemarmiesse commented:
@SriRangaTarun Please merge master into your branches to make sure that it's compatible with tf.keras

SriRangaTarun commented:
@RaphaelMeudec and @gabrieldemarmiesse I am closing this PR and have created an [alias here](https://github.com/keras-team/keras-contrib/pull/459). There, I have merged master into my branch.
Any other changes ?
